### DISCUSSION TITLE
How do you conceal the space between nodes?

### DISCUSSION THREAD

**Discussion by ColinKennedy** [neovim/neovim#25748] [Q&A] [answered]
I'm after a specific behavior and am looking for ways to do it. I think I should be looking into conceal and folds but at this point I'm trying to figure out what's possible.

With the introduction to [tree-sitter-objdump](https://github.com/ColinKennedy/tree-sitter-objdump) into [nvim-treesitter](github.com/nvim-treesitter/nvim-treesitter), it's now possible to parse objdump output separately from its generated assembly. I want to be able to "toggle" between the raw objdump text and just the assembly code, as if it was a written assembly file. For example:

```
0000000000000000 <engine::world::World::createEntity()> (File Offset: 0x70):
   0:	48 89 f8                                              	mov    rax,rdi  # 130 <big_call_signature+0x130> (File Offset: 0x40c)
   3:	48 8b 4f 40                                           	mov    rcx,QWORD PTR [rdi+0x40]   # 0x4c1788
```

Would conceal/hide as

```
mov    rax,rdi
mov    rcx,QWORD PTR [rdi+0x40]
```

And because it's a conceal, I keep all of the highlight and original objdump context for other fancy features that need it. It's an important piece of what I'm trying to do, on top of being able to toggle between them.

Hiding the `0000000000000000 ...` line should be possible, using folds. But I'm struggling to figure out how to condense the Assembly lines. The tree-sitter contains no named/unamed nodes between the `48 89 f8` part and the `mov    rax,rdi` part. To tree-sitter, it's just whitespace. Since there's no node, I can't make a tree-sitter query to conceal it. How would I collapse that text?

**Comment by clason** [comment, marked as answer]
Duplicate of https://github.com/neovim/neovim/issues/25515

**Reply by clason** [reply]
And I feel that this is just going too far -- Neovim is an editor, not a pretty-printer. Use a custom tool for that.

**Reply by ColinKennedy** [reply]
I don't believe that it is. The titles are similar but the user is asking for concealing between two nodes (inclusive) and I'm asking for between two nodes (exclusive). Vim already has `:h syn-concealends` for conceal ranges so I don't see why a tree-sitter based conceal range is considered off-topic. Could you please explain your opinion?

**Reply by clason** [reply]
Well, exclusive vs. inclusive is a minor detail. The issue is that treesitter can handle nodes individually, period, but not the space between nodes. It's just not built for that.

So this would require a whole new tree-sitter API on top of what we have, together with custom queries. And that's simply a tough sell.

**Reply by ColinKennedy** [reply]
Then the use case creates friction up against tree-sitter and not necessarily Neovim. I'm in a weird spot at the moment because I have a vision of something I want to create but my situation is either ...

- Add a bunch of unnamed nodes to the tree-sitter-objdump grammar so I'll have something to query-and-conceal
- Write an external tool from scratch / find another tool than Neovim

For reasons that I think align super well with Neovim, I'd like not to go full-external. However this rough spot is definitely hard to solve, haha.

Things would be much simpler if I could conceal hidden tree-sitter nodes but when I tried it, it doesn't work. Either because Neovim doesn't support it or tree-sitter or possibly both. You can design conceal queries on nodes and unnamed nodes but not hidden nodes. My bet is on it being "tree-sitter hidden nodes aren't in the tree and that's why you can't query them" but I don't know.

**Reply by clason** [reply]
> For reasons that I think align super well with Neovim,

From what I've heard so far, I actually don't think so. I know a "one stop shop" sounds convenient, but scope creep is the nemesis of open source projects, so we have to be vigilant.

And, yes, "hidden nodes" are hidden. If you want to do something (anything) with them, they shouldn't be hidden.

**Reply by ColinKennedy** [reply]
I get that you wouldn't want niche, special interests from causing problems for Neovim as an OSS but I don't see a reason to discourage a plugin that hasn't been explained fully and no one has seen. Anyway, my question has been answered so I'll mark this as answered. Thank you for taking the time to chat!

**Reply by clason** [reply]
Not discouraging a plugin; just saying that it's unlikely that Neovim would add a completely new API just for it, from what you explained so far ;)
